### PR TITLE
remove extra node in prop path

### PR DESCRIPTION
**Pull request by EdColeman**
The original rework of the prop store was using a separate ZooKeeper node under the "original"  configuration node.  In ZooKeeper the structure looked like:
 
```
/accumulo/[instance id]/config/encoded_props  <-- system props
/accumulo/[instance id]/namespaces/[nid]/conf/encoded_props  <-- namespace props
/accumulo/[instance id]/tables/[tid]/conf/encoded_props  <-- table props

```

This removed that extra node an stores the configuration directly in the configuration node.  The structure now looks like:

```
/accumulo/[instance id]/config  <-- system props
/accumulo/[instance id]/namespaces/[nid]/config  <-- namespace props
/accumulo/[instance id]/tables/[tid]/config  <-- table props
```

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -695,6 +695,11 @@
       <artifactId>spotbugs-annotations</artifactId>
       <optional>true</optional>
     </dependency>
+    <dependency>
+      <groupId>org.apache.logging.log4j</groupId>
+      <artifactId>log4j-core</artifactId>
+      <scope>provided</scope>
+    </dependency>
```
Comment (ctubbsii) [inline pom.xml:702]: We don't need to use provided scope. That is intended for shipping packages that depend on stuff provided by their containerized environment, like a Maven plugin expected to run inside Maven, or a WAR intended to run inside an application server.

Instead, we need to explicitly depend on this as a `test` dependency in places where we need it, or as an optional runtime dependency. For us, the runtime dependency doesn't really matter because we only need it for the distribution tarball.

```suggestion
```

**Review by ctubbsii — COMMENTED** [review]
I do think it would be nice if system scope used the same node name (config or conf, consistently). We could also fix that here before the release.

**Comment by ctubbsii** [comment]
This would fix #2690

File: pom.xml (hunk 1 above)
Comment (EdColeman) [reply pom.xml:702]: This should be removed when #3000 is merged.

**Comment by EdColeman** [comment]
Changing the node name so that they match would require additional changes that would include removing the legacy node.  It was easier now to keep the existing convention.  I would suggest a follow-on PR.  It is mainly going to be changes in the ConfigurationTransform code.

A higher priority right now should be working to trace down the concurrency issues with multiple property updates - once those are resolved, this could be revisited.

**Comment by cshannon** [comment]
> Changing the node name so that they match would require additional changes that would include removing the legacy node. It was easier now to keep the existing convention. I would suggest a follow-on PR. It is mainly going to be changes in the ConfigurationTransform code.
> 
> A higher priority right now should be working to trace down the concurrency issues with multiple property updates - once those are resolved, this could be revisited.

Tests pass with the latest in PR #2967, See my comment here: https://github.com/apache/accumulo/pull/2967#issuecomment-1271706513

**Comment by cshannon** [comment]
This change doesn't make sense to me to merge in now as it seems too late to be making a change like this with the 2.1.0 release being imminent. I would think it would be better to wait until the next version for this change and at the same time we can also rename system scope to be consistent with the others.

**Comment by cshannon** [comment]
Also if the reason to do it now would be that changing it later requires doing some upgrading/migration to the property store then that could just be part of the migration that needs to be done anyways if we rename the system level node from config to conf.

**Comment by cshannon** [comment]
What's the status of this PR now? Is there anything else to do here or is it ready?

**Comment by EdColeman** [comment]
I have an update forthcoming that uses the same node name (config) instead of config for system and conf for namespace and tables - going through the final clean-up now.  Because of merges, the push will likely be a force push.

**Comment by EdColeman** [comment]
This now contains the changes needed to use config instead of conf as the configuration node name for tables and namespaces (system already used config)

**Review by cshannon — CHANGES_REQUESTED** [review]
I think this mostly looks good but I think a couple things could be changed which I left comments for. 

I also think some of the testing should be improved a bit specifically for `ConfigTransformer` and `ConfigProperyUpgrader`. Since this is upgrading properties I think it should be well tested as it wouldn't be great if the upgrade logic failed and broke an existing install when a user upgraded.

1. `ConfigProperyUpgraderTest` was created but just uses a mock `ConfigTransformer`. It also doesn't test much other than the normal path. There's some edge cases that could be tested I think to make sure it handles errors. For example test what happens if the legacy path is bad, etc.
2.For `ConfigTransformer`, there are a couple integration tests (only `ConfigTransformerIT`) but not many. I think it might be useful to create a test specifically for `ConfigTransformer` (just like you did for `ConfigPropertyUpgrader`) to test the new behavior (using mocks, etc) and run through the different edge cases and scenarios to verify it works.
3.Lastly, I think both `ConfigPropertyUpgraderIT` and `ConfigTransformerIT` could use some extra tests to exercise the new behavior in more scenarios.

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java [hunk 2]
```diff
@@ -438,6 +443,9 @@ private VersionedProperties readPropsFromZk(PropStoreKey<?> propStoreKey)
     try {
       Stat stat = new Stat();
       byte[] bytes = zrw.getData(propStoreKey.getPath(), stat);
+      if (bytes.length == 0) {
```
Comment (cshannon) [inline server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java:446]: Same comment/question about bytes being null as above.

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/PropStoreKey.java [hunk 3]
```diff
@@ -90,24 +85,33 @@ protected PropStoreKey(final InstanceId instanceId, final String path, final ID_
...
       return TablePropKey.of(instanceId, TableId.of(tokens[ID_TOKEN_POSITION]));
     }
-    if (tokens[TYPE_TOKEN_POSITION].equals(NAMESPACE_NODE_NAME)) {
+
+    if (tokens[TYPE_TOKEN_POSITION].equals(NAMESPACE_NODE_NAME) && nodeName.equals(ZCONFIG)) {
       return NamespacePropKey.of(instanceId, NamespaceId.of(tokens[ID_TOKEN_POSITION]));
     }
-    return SystemPropKey.of(instanceId);
+
+    if (nodeName.equals(ZCONFIG)) {
```
[the remark this answers is not in the document]
Comment (EdColeman) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/PropStoreKey.java:109]: They are ordered in the most likely occurrence  - in general the number of tables > number of namespaces > system so it seemed rational to order the check so that what would (hopefully) be more common to be checked first.

File: server/base/src/main/java/org/apache/accumulo/server/conf/util/ConfigTransformer.java [hunk 4]
```diff
@@ -152,18 +152,15 @@ VersionedProperties transform(final PropStoreKey<?> propStoreKey, final Transfor
...
-        results = ZooPropStore.readFromZk(propStoreKey, propStoreWatcher, zrw);
-        if (results != null) {
-          return results;
-        }
+        return writeNode(propStoreKey, Map.of());
       }
 
       upgradeNodes = convertDeprecatedProps(propStoreKey, upgradeNodes);
-
+      // todo - here
```
[the remark this answers is not in the document]
Comment (EdColeman) [reply server/base/src/main/java/org/apache/accumulo/server/conf/util/ConfigTransformer.java:163]: No

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/PropStoreKey.java [hunk 5]
```diff
@@ -90,24 +85,33 @@ protected PropStoreKey(final InstanceId instanceId, final String path, final ID_
...
       return null;
     }
-    if (tokens.length < 1 || !tokens[tokens.length - 1].equals(PROP_NODE_NAME)) {
-      // without tokens or it does not end with PROP_NAME_NAME
-      return null;
-    }
-    if (tokens[TYPE_TOKEN_POSITION].equals(TABLES_NODE_NAME)) {
+
+    String nodeName = "/" + tokens[tokens.length - 1];
+    if (tokens[TYPE_TOKEN_POSITION].equals(TABLES_NODE_NAME) && nodeName.equals(ZCONFIG)) {
```
[the remark this answers is not in the document]
Comment (EdColeman) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/PropStoreKey.java]: addressed in ac8d0b8da8

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropLoader.java [hunk 6]
```diff
@@ -63,6 +63,9 @@ public ZooPropLoader(final ZooReaderWriter zrw, final VersionedPropCodec propCod
 
       Stat stat = new Stat();
       byte[] bytes = zrw.getData(propStoreKey.getPath(), propStoreWatcher, stat);
+      if (bytes.length == 0) {
```
[the remark this answers is not in the document]
Comment (EdColeman) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropLoader.java]: Addressed in ac8d0b8da8 - in this case, the ZooKeeper stat is also being used, so that should be another reliable check for the data length.  The NoNodeException likely was sufficient - but checking the stat does not seem to hurt.

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java [hunk 7]
```diff
@@ -222,6 +223,10 @@ public void create(PropStoreKey<?> propStoreKey, Map<String,String> props) {
     try {
       Stat stat = new Stat();
       byte[] bytes = zooReader.getData(propStoreKey.getPath(), watcher, stat);
+      if (bytes.length == 0) {
```
[the remark this answers is not in the document]
Comment (EdColeman) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java]: Addressed in ac8d0b8da8 and see previous comment

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java (hunk 2 above)
Comment (EdColeman) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java]: Addressed in ac8d0b8da8 and see previous comment

**Review by cshannon — APPROVED** [review]

**Comment by cshannon** [comment]
LGTM now

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/PropStoreKey.java [hunk 8]
```diff
@@ -89,25 +89,44 @@ protected PropStoreKey(final InstanceId instanceId, final String path, final ID_
...
   public static @Nullable PropStoreKey<?> fromPath(final String path) {
     String[] tokens = path.split("/");
+    if (tokens.length < 1) {
+      return null;
+    }
 
     InstanceId instanceId;
     try {
       instanceId = InstanceId.of(tokens[IID_TOKEN_POSITION]);
+
```
Comment (dlmarion) [inline server/base/src/main/java/org/apache/accumulo/server/conf/store/PropStoreKey.java:99]: Alternatively, you could remove the try/catch here and check that tokens.length equals EXPECTED_CONFIG_LEN or EXPECTED_SYS_CONFIG_LEN.

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/PropStoreKey.java [hunk 9]
```diff
@@ -89,25 +89,44 @@ protected PropStoreKey(final InstanceId instanceId, final String path, final ID_
...
+      // needs to start with /accumulo/[instanceId]
+      if (!path.startsWith(ZooUtil.getRoot(instanceId))) {
+        log.warn(
+            "Path '{}' is invalid for a property cache key, expected to start with /accumulo/[instance_id]",
```
Comment (dlmarion) [inline server/base/src/main/java/org/apache/accumulo/server/conf/store/PropStoreKey.java:103]: Is there a reason not to put the actual instanceId in the warning message?

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/PropStoreKey.java (hunk 8 above)
Comment (EdColeman) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/PropStoreKey.java]: addressed in b9ef24d03b

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/PropStoreKey.java (hunk 9 above)
Comment (EdColeman) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/PropStoreKey.java]: addressed in b9ef24d03b

**Review by dlmarion — APPROVED** [review]
LGTM
